cores/arduino/soc_i2c.c

```diff
@@ -0,0 +1,326 @@
+	      return I2C_ERROR_ADDRESS_NOACK; // NACK on transmit of address
+            }
+	    else if (soc_i2c_err_source[controller_id] & I2C_ABRT_TXDATA_NOACK) {
+	      return I2C_ERROR_DATA_NOACK; // NACK on transmit of data
+            } else {
+	      return I2C_ERROR_OTHER; // other error
+            }
+        }
+        if (soc_i2c_master_tx_complete[controller_id]) {
+            return I2C_OK;
+        }
+        delayMicroseconds(10);
+    }
+    return I2C_TIMEOUT;
+}
+
+static int soc_i2c_wait_dev_ready(SOC_I2C_CONTROLLER controller_id,
+                                  bool no_stop)
+{
+    uint64_t timeout = TIMEOUT_MS * 200;
+    int ret = 0;
+    while (timeout--) {
+        ret = soc_i2c_status(controller_id, no_stop);
+        if (ret == I2C_OK) {
+            return I2C_OK;
+        }
+        if (ret == I2C_BUSY) {
+            delayMicroseconds(10);
+        }
+    }
+    return I2C_TIMEOUT - ret;
+}
+
+int soc_i2c_open_adapter(SOC_I2C_CONTROLLER controller_id, uint32_t address, int i2c_speed, int i2c_addr_mode)
+{
+    int ret = 0;
+
+    if (controller_id == SOC_I2C_0) {
+      SET_PIN_MODE(20, I2C_MUX_MODE);
+      SET_PIN_MODE(21, I2C_MUX_MODE);
+
+      SET_PIN_PULLUP(20, 1);
+      SET_PIN_PULLUP(21, 1);
+    } else {
+      SET_PIN_MODE(22, I2C_MUX_MODE);
+      SET_PIN_MODE(23, I2C_MUX_MODE);
+
+      SET_PIN_PULLUP(22, 1);
+      SET_PIN_PULLUP(23, 1);
+    }
+
+    i2c_cfg_data_t i2c_cfg;
+    memset(&i2c_cfg, 0, sizeof(i2c_cfg_data_t));
+
+    i2c_cfg.speed = i2c_speed;
+    i2c_cfg.addressing_mode = i2c_addr_mode;
+    if (address) {
+        i2c_cfg.mode_type = I2C_SLAVE;
+	if (controller_id == SOC_I2C_0) {
+	  i2c_cfg.cb_err = soc_i2c0_err_callback;
+	  i2c_cfg.cb_rx = soc_i2c0_slave_rx_callback;
+	  i2c_cfg.cb_tx = soc_i2c0_slave_tx_callback;
+	} else {
+	  i2c_cfg.cb_err = soc_i2c1_err_callback;
+	  i2c_cfg.cb_rx = soc_i2c1_slave_rx_callback;
+	  i2c_cfg.cb_tx = soc_i2c1_slave_tx_callback;
+	}
+    } else {
+        i2c_cfg.mode_type = I2C_MASTER;
+
+	if (controller_id == SOC_I2C_0) {
+	  i2c_cfg.cb_tx = soc_i2c0_master_tx_callback;
+	  i2c_cfg.cb_rx = soc_i2c0_master_rx_callback;
+	  i2c_cfg.cb_err = soc_i2c0_err_callback;
+	} else {
+	  i2c_cfg.cb_tx = soc_i2c1_master_tx_callback;
+	  i2c_cfg.cb_rx = soc_i2c1_master_rx_callback;
+	  i2c_cfg.cb_err = soc_i2c1_err_callback;
+	}
+        soc_i2c_master_tx_complete[controller_id] = 0;
+        soc_i2c_master_rx_complete[controller_id] = 0;
+    }
+    i2c_cfg.slave_adr = address;
+    soc_i2c_err_detect[controller_id] = 0;
+    soc_i2c_err_source[controller_id] = 0;
+
+    soc_i2c_set_config(controller_id, &i2c_cfg);
+    soc_i2c_clock_enable(controller_id);
+
+    ret = soc_i2c_wait_dev_ready(controller_id, false);
+
+    return ret;
+}
+
+void soc_i2c_close_adapter(SOC_I2C_CONTROLLER controller_id)
+{
+    soc_i2c_deconfig(controller_id);
+    soc_i2c_clock_disable(controller_id);
+
+    SET_PIN_MODE(20, GPIO_MUX_MODE);
```

Comment:
it should according to the I2C controller id to set different pins, for I2C0, the pins are 20 and 21, for I2C1 the pins are 22 and 23